In [2]:
import tensorflow as tf

In [2]:
tf.__version__

'1.15.0'

## 오차 수정하기 : 경사하강법
-  미분의 기울기를 이용해 오차를 비교해 가장 작은 방향으로 이동시킴
-  (x에대한)미분값이 0인 지점/기울기가 0이 되는(오차값)을 찾는 것
- 오차와 기울기 a로 이루어진 함수는 이차함수
- a1에서 기울기를 구하고 반대의 a2의 기울기를 구할때 , 기울기가 0이되는 m의 값을 찾는다.
- 학습률 runnning rate

In [4]:
x = tf.constant([[1.0,2.0,3.0]])  # 1*3 행렬만들기
w = tf.constant([[2.0],[2.0],[2.0]]) # 3*1 행렬만들기
y = tf.matmul(x,w)  # 행렬 곱

print(x.get_shape()) 
print(w.get_shape())
print(y.get_shape())
print(y)  # 연산의 결과가 아닌, 연산을 할 수 있는 tensor

sess = tf.Session()    # 실행시킬 수 있는 기계, session에 tensor를 집어넣고 실행시킴. 2.0버전에서는 없어짐
init = tf.global_variables_initializer()
sess.run(init)
result = sess.run(y)
print(result)

(1, 3)
(3, 1)
(1, 1)
Tensor("MatMul_1:0", shape=(1, 1), dtype=float32)
[[12.]]


In [12]:

 a = tf.constant(7.0,name = 'data1')  
 b = tf.constant(3.0,name = 'data2')  # tensor는 변수처럼 사용하는 개체
 c = tf.constant(2.0,name = 'data3')
    
v = a * b / c

# with as 구문을 사용해, 자동으로 close
with tf.Session() as sess: 
    # tensor 초기화  # 변수의 초기화
    sess.run(tf.global_variables_initializer()) 
    # v tensor 가동
    print(sess.run(v)) 
    
print(type(v))

10.5
<class 'tensorflow.python.framework.ops.Tensor'>


tensorflow module을 import함

이때 Defalut Graph가 생김

In [5]:
a = tf.constant(7.0,name = 'data1')  
# a (엣지) = operation
b = tf.constant(3.0,name = 'data2')  # tensor는 변수처럼 사용하는 개체
c = tf.constant(2.0,name = 'data3')

In [6]:
# 그래프 확인
# kernel -> restart 로 위에 Session 닫기
# const = 상수

tf.get_default_graph().get_operations()  

[<tf.Operation 'data1' type=Const>,
 <tf.Operation 'data2' type=Const>,
 <tf.Operation 'data3' type=Const>]

In [7]:
# a의 값을 확인
# operation이 실행되기 전까지 텐서 a의 값이 결정되지 않은 상태
print(a)
# a에 연결된 operation을 확인
print(a.op)

Tensor("data1:0", shape=(), dtype=float32)
name: "data1"
op: "Const"
attr {
  key: "dtype"
  value {
    type: DT_FLOAT
  }
}
attr {
  key: "value"
  value {
    tensor {
      dtype: DT_FLOAT
      tensor_shape {
      }
      float_val: 7.0
    }
  }
}



In [8]:
print(a.op.outputs) #operation a의 출력에 접근하여 확인

[<tf.Tensor 'data1:0' shape=() dtype=float32>]


In [12]:
print(a is a.op.outputs[0])  
# a는 a.op.ouputs을 의미한다.
# a는 클래스의 객체의 성격을 띈다.
# float_val

True


In [13]:
v = a * b / c

In [14]:
tf.get_default_graph().get_operations()
#  마지막에 텐서플로우의 연산 노드가 자동으로 추가됨
# node에 이름을 붙이지 않았으므로 default 이름인 mul, truediv를 가짐

[<tf.Operation 'data1' type=Const>,
 <tf.Operation 'data2' type=Const>,
 <tf.Operation 'data3' type=Const>,
 <tf.Operation 'mul' type=Mul>,
 <tf.Operation 'truediv' type=RealDiv>]

In [17]:
print('v.op.inputs=',v.op.inputs)
# 리스트 객체의 id가 출력

v.op.inputs= <tensorflow.python.framework.ops.Operation._InputList object at 0x000001DC3FBB05C8>


In [18]:
print('v.op.outputs=',v.op.outputs)
# 텐서 v의 operation의 출력을 확인

v.op.outputs= [<tf.Tensor 'truediv:0' shape=() dtype=float32>]


텐서는 node와 edge를 가짐

tensorflow a = 이라고 만들면

뒤에는 edge

a라는 텐서는 안에 여러가지 정보를 마치 클래스의 정보처럼 가지고 있음.

In [20]:
# v가 결과
# 그래프의 정보를 가지고있는 v도 객체 . tensor가 된다.
print(v.op)

name: "truediv"
op: "RealDiv"
input: "mul"
input: "data3"
attr {
  key: "T"
  value {
    type: DT_FLOAT
  }
}



텐서는 다차원 배열을 의미하는데, Rank, Shape, Type의 3가지 구성요소를 가지고 있다.
-  rank 0 텐서는 shpae[] 스칼라 ex,3
-  rank 1 텐서는 벡터 ex) [1,2,3]
-  rank 2 텐서는 행렬 ex [[1,2,3],[4,5,6]]
-  그 이외는 n차원 텐서
-  rank는 텐서의 차원을 의미한다.
-  shape란 텐서가 몇개의 행과 열을 가지는 지를 의미
-  텐서의 type이란 텐서가 담을 수 있는 데이터의 타입을 의미한다.

텐서플로우 자료형

-  tf.constant 변하지 않는 수를 지정하는 자료형
-  tf.Placeholder 먼저 만들고 값을 나중에 지정해주는 자료형
-  x  = tf.Placeholder(tf.float32), 변수처럼 만들어놓고 필요할 때 객체에 필요한 데이터를 넣는다.
-  tf.Variable 조건에 따라 값을 바꿔야 하는 경우 사용하는 자료형, 결과값을 설정할 때
-  Placeholder와 Variable의 차이, 만들어 놓고 나중에 값이 할당된다는 기능이 같기 때문에 호환가능,,?


텐서플로우의 상수 연산

In [27]:
a = tf.constant(5)
b = tf.constant(2)
c = tf.constant(3)
d = tf.multiply(a,b)   # a * b
e = tf.add(c,d)        # c + d
f = tf.subtract(d,e)   # d - e

In [28]:
print(a)
print(b)
print(c)
print(d)
print(e)
print(f)

Tensor("Const_7:0", shape=(), dtype=int32)
Tensor("Const_8:0", shape=(), dtype=int32)
Tensor("Const_9:0", shape=(), dtype=int32)
Tensor("Mul_2:0", shape=(), dtype=int32)
Tensor("Add_1:0", shape=(), dtype=int32)
Tensor("Sub:0", shape=(), dtype=int32)


-  위의 상수들은 모두 텐서
-  텐서를 만들었을 뿐, 실행을 하지않음
-  그래프를 정의했을 뿌닝지, 실제로 연산이 일어난 것을 아님
-  연산을 수행하기 위해서는 데이터를 넣어서 흐름이 이루어지게 하여야 함. (이러한 동작을 Session이 수행)

In [29]:
sess = tf.Session()
outs = sess.run(f)
sess.close()
print('out = ',outs)

out =  -3


### 그래프의 생성과 실행

In [31]:
import tensorflow as tf

# 문자열이 들어있는 node생성
hello = tf.constant('Hello, Tensorflow!')

# TF 세션 시작
sess = tf.Session()

# 특정 노드 세션 run
print(sess.run(hello))

b'Hello, Tensorflow!'


변수의 초기화
-  variable.이나 placeholder를 사용할때는 초기화?

-  placeholder로 만들어준것은 feed.dict를 사용해서 값을 넣어준다.

In [36]:
p_holder1 = tf.placeholder(dtype=tf.float32) # 변수 만들기 # placeholder는 초기에 값을 넣지 못함
p_holder2 = tf.placeholder(dtype=tf.float32)
p_holder3 = tf.placeholder(dtype=tf.float32)

val1 = 5
val2 = 10
val3 = 3

ret_val = p_holder1 * p_holder2 + p_holder3

feed_dict_data = {p_holder1:val1, p_holder2:val2,p_holder3:val3}

# placeholder로 만들어준것은 feed.dict를 사용해서 값을 넣어준다.
result = sess.run(ret_val, feed_dict=feed_dict_data)

print(result)

53.0


### placeholder & feed_dict

In [39]:
I = tf.placeholder(tf.float32,[2,4])
i = [[1,2,3,4],[5,6,7,8]]  #데이터가 초기화 되어있는 배열 만들기

W = tf.placeholder(tf.float32,[4,2])
w = [[1,1],[2,2],[3,3],[4,4]]

#input 데이터와 Weight를 행렬곱 수행
node = tf.matmul(I,W)  

sess = tf.Session()  #세션만들기

print(sess.run(node,feed_dict={I:i,W:w}))  

sess.close

[[30. 30.]
 [70. 70.]]


<bound method BaseSession.close of <tensorflow.python.client.session.Session object at 0x000001DC3A47F8C8>>

### placeholder & Variable

In [43]:

I = tf.Variable([[1,2,3,4],[5,6,7,8]],tf.float32)

W = tf.Variable([[1,1],[2,2],[3,3],[4,4]],tf.float32)

# input 데이터와 weight 행렬곱 수행
node = tf.matmul(I,W)

sess = tf.Session()
sess.run(tf.global_variables_initializer())

print(sess.run(node))

sess.close()



[[30 30]
 [70 70]]


### 변수를 재할당하는 방법

변수 초기화
-  변수초기화는 모델의 다른 연산을 실행하기 전에 반드시 명시적으로 실행해야 한다.
-  모든 변수를 초기화 하는 연산을 모델 사용 전에 실핸한다.
-  변수 초기화를 위한 작업(op)를 추가하기 위해 tf.global_variables_intializer()
-  모델을 모두 만들고 세션에 올린 후 이 작업을 실핸한다.
-  Variable은 반드시 초기화
-  tensor에서 만드는 변수는 행렬의 형태를 띄기 때문에 메모리에 할당되었다가 지워진 변수를 초기화해주고 실행해야함.


In [51]:
x = tf. Variable(5, dtype = tf.int32)
print(x)

x = tf.assign(x,7)
z = tf.initialize_all_variables()     # 모든 변수 초기화
sess = tf.Session()   # 세션에서 그래프 시작
sess.run(z)    #  텐서 z를 계산
y = sess.run(x)
print(y)

sess.close()

<tf.Variable 'Variable_15:0' shape=() dtype=int32_ref>
7


### 코딩으로 확인하는 경사 확인법

In [52]:
data = [[2,81],[4,93],[6,91],[8,97]]

# x,y의 데이터 값
x_data = [x_row[0] for x_row in data]
y_data = [y_row[1] for y_row in data]

# 기울기 a와 y절편 b의 값을 임의로 정한다.
#단, 기울기의 범위는 0~10 사이이며, y절편은 0~100 사이에서 변하게 한다.
# [1] 0에서 10사이에서 임의의 수 1개

a = tf.Variable(tf.random_uniform([1],0,10, dtype = tf.float64,seed = 0)) # seed 동일한 패턴의 난수
b = tf.Variable(tf.random_uniform([1],0,100, dtype = tf.float64,seed = 0))

# y에 대한 일차 방정식 ax+b의 식을 세운다.
# a,b가 tensor 이니까 y또한 tensor로 만들어짐  => session 을 만들어서 계산을 해야한다.

y = a* x_data + b

-  tf.reduce_mean 사용된 data들 각각에서 계산된 합의 평균을 구하는 함수
-  tf.square 제곱함수
-  tf.sqrt 제곱근합수
-  tf.train.GradientDescentOptimizer  rmse를 최소화할 수 있는 경사하강법 

learning rate란?
- 기울기의 부호를 바꿔 이동시킬 때 적절한 거리를 찾지 못해 너무 멀리 이동시키면 a 값이 한 점으로 모이지 않고 위로 치솟아 버림
- 어느만큼 이동시킬지를 정해주는 것.
- 학습률을 너무 크게 잡으면 한 점으로 수렴하지않고 발산한다.
- 데이터에 따라 가변될 수 있기 때문에 사용자가 직접 지정해서 값을 넣어준다.
- 이동하는 거리를 결정
- 한 번의 학습으로 매개변수 값을 얼마나 갱신하는가를 정하는 것
- 초 매개변수????????????????????
- 0.1 또는 0.01전후로 정해 학습을 시도한다.
- 학습률이너무 작거나 크면 학습이 제대로 이루어지지 않는다.
- 너무 작으면 연산을 너무 많이하고(수렴이 늦어지고, 지역최소점에 걸린다), 
- 크면 위로 발산한다. ( 손실함수 값이 줄어들지 않고 오히려 더 커지는 오버슈팅) 

In [59]:
#텐서플로 RMSE 함수
# 평균제곱근오차
# rmse도 tensor로 계산됨
rmse = tf.sqrt(tf.reduce_mean(tf.square( y- y_data)))

# 학습률 값
learning_rate = 0.1

# RMSE 값을 최소로 하는 값 찾기
gradient_decent = tf.train.GradientDescentOptimizer(learning_rate).minimize(rmse)

# optimizer = tf.train.GradientDescentOptimizer(learning_rate = 0.1)
# train = optimizer.minimize(rmse)


#텐서플로를 이용한 학습
#세션 실행
with tf.Session() as sess:  
    
    #변수 초기화
    sess.run(tf.global_variables_initializer())
    
    # 2001번 실행 (0번째를 포함하므로)
    for step in range(2001):
        # rmse를 최소화하는 값을 내리는 결과를 2001번 반복
        # 그래프 run
        # sess.run('그래프명')
        sess.run(gradient_decent)
        
        #100번마다 결과 출력
        if step %100 == 0:
            # Epoch 몇 번 반복했는지? (= step)
            # RMSE 평균제곱근 오차값
            print('Epoch: %.f, RMSE = %.04f, 기울기 a = %.4f, y절편 b = %.4f'%
                 (step,sess.run(rmse),sess.run(a),sess.run(b)))

Epoch: 0, RMSE = 30.2139, 기울기 a = 7.5235, y절편 b = 80.5984
Epoch: 100, RMSE = 2.8860, 기울기 a = 2.2299, y절편 b = 79.4181
Epoch: 200, RMSE = 2.8826, 기울기 a = 2.2601, y절편 b = 79.2379
Epoch: 300, RMSE = 2.8815, 기울기 a = 2.2773, y절편 b = 79.1353
Epoch: 400, RMSE = 2.8811, 기울기 a = 2.2871, y절편 b = 79.0770
Epoch: 500, RMSE = 2.8810, 기울기 a = 2.2927, y절편 b = 79.0438
Epoch: 600, RMSE = 2.8810, 기울기 a = 2.2958, y절편 b = 79.0249
Epoch: 700, RMSE = 2.8810, 기울기 a = 2.2976, y절편 b = 79.0142
Epoch: 800, RMSE = 2.8810, 기울기 a = 2.2987, y절편 b = 79.0081
Epoch: 900, RMSE = 2.8810, 기울기 a = 2.2992, y절편 b = 79.0046
Epoch: 1000, RMSE = 2.8810, 기울기 a = 2.2996, y절편 b = 79.0026
Epoch: 1100, RMSE = 2.8810, 기울기 a = 2.2998, y절편 b = 79.0015
Epoch: 1200, RMSE = 2.8810, 기울기 a = 2.2999, y절편 b = 79.0008
Epoch: 1300, RMSE = 2.8810, 기울기 a = 2.2999, y절편 b = 79.0005
Epoch: 1400, RMSE = 2.8810, 기울기 a = 2.3000, y절편 b = 79.0003
Epoch: 1500, RMSE = 2.8810, 기울기 a = 2.3000, y절편 b = 79.0002
Epoch: 1600, RMSE = 2.8810, 기울기 a = 2.3000, y절편 b =

=> a가 2.3에 수렴하고 b가 79에 수렴함
-  y = 2.3x + 79


- 변수 할당 -> 그래프 생성 (아직 실행 X)
- 세션에서 Run을 해주어야하기 때문에 (run할 때 사용되기 떄문에 run하기 전에 변수 초기화)